### PR TITLE
blob: Add support for ListOptions.Delimiter

### LINKED ISSUE #541: blob: add support for Delimiters as part of List
Please use a title starting with the name of the affected package, or \"all\",
followed by a colon, followed by a short summary of the feature request.
Example: `blob/gcsblob: add support for more blobbing`.

### Is your feature request related to a problem? Please describe.

Related to #241.

### Describe the solution you'd like

AWS and GCS both support a `Delimiter` for directory-like behavior.

The delimiter allows you to treat the key space semi-hierarchically. For example, if you had the following blobs:

- `foo`
- `bar`
- `baz/quux`
- `baz/fido`

And you did a list with empty prefix with the slash delimiter, you would get back `foo`, `bar`, and `baz/` (the API makes some indication that `baz/` is a dummy entry).

### PR DESCRIPTION
**Pull request by vangent**
Fixes #541.

Some known followups in terms of adding comments and tests are already tracked in #577.

### CODE REVIEW

File: blob/fileblob/fileblob.go [hunk 1]
```diff
@@ -16,10 +16,12 @@
 // filesystem. This should not be used for production: it is intended for local
 // development.
 //
-// Blob names must only contain alphanumeric characters, slashes, periods,
-// spaces, underscores, and dashes. Repeated slashes, a leading "./" or "../",
-// or the sequence "/./" is not permitted. This is to ensure that blob names map
-// cleanly onto files underneath a directory.
+// Blob names are escaped using url.PathEscape before writing them
```
Comment (zombiezen) [inline blob/fileblob/fileblob.go]: Let's do this as a separate change: there's enough other business going on in this PR.

File: blob/drivertest/drivertest.go [hunk 2]
```diff
@@ -319,6 +321,219 @@ func testList(t *testing.T, newHarness HarnessMaker) {
...
+	if err != nil {
+		return nil, err
+	}
+	var retval []*listResult
+	for {
+		obj, err := iter.Next(ctx)
+		if err != nil {
+			return nil, err
+		}
+		if obj == nil {
```
Comment (jba) [inline blob/drivertest/drivertest.go]: I think when we adopt an error sentinel for Next, this case will go away (become impossible).

File: blob/example_test.go [hunk 3]
```diff
@@ -208,6 +208,76 @@ func ExampleBucket_List() {
...
+			// Directory.
+			fmt.Printf("%s%s\n", indent, obj.Prefix)
+			List(ctx, b, obj.Prefix, indent+"  ")
+		} else {
+			fmt.Printf("%s%s\n", indent, obj.Key)
+		}
+	}
+}
+
+func ExampleBucket_List_withdelimiter() {
```
[the remark this answers is not in the document]
Comment (vangent) [reply blob/example_test.go]: Done.

File: blob/example_test.go [hunk 4]
```diff
@@ -208,6 +208,76 @@ func ExampleBucket_List() {
...
+// List lists files in b starting with prefix. It uses the delimiter "/",
+// and recurses into "directories", adding 2 spaces to indent each time.
+// The result is an indented listing like this:
+// foo/
+//   a.txt
+//   b.txt
+// bar/
+//   barsubdir/
+//     c.txt
+func List(ctx context.Context, b *blob.Bucket, prefix, indent string) {
```
[the remark this answers is not in the document]
Comment (vangent) [reply blob/example_test.go]: Done.

File: blob/drivertest/drivertest.go (hunk 2 above)
Comment (vangent) [reply blob/drivertest/drivertest.go]: Agree.

File: blob/fileblob/fileblob.go (hunk 1 above)
Comment (vangent) [reply blob/fileblob/fileblob.go]: Done, sent PR #588.

**Comment by vangent** [comment]
Grabbing this until #588 is submitted and merged.

**Comment by vangent** [comment]
Changes merged, PTAL.

File: blob/driver/driver.go [hunk 5]
```diff
@@ -139,6 +150,11 @@ type ListObject struct {
 	ModTime time.Time
 	// Size is the size of the object in bytes.
 	Size int64
+	// IsDir indicates that this result represents a "directory" in the
+	// hierarchical namespace, ending in ListOptions.Delimiter. Key can be
+	// passed as ListOptions.Prefix to list items in the "directory".
+	// Fields other than Key and IsDir will not be set.
```
Comment (zombiezen) [inline blob/driver/driver.go]:
```suggestion
	// Fields other than Key and IsDir will not be set if IsDir is true.
```

File: blob/example_test.go [hunk 6]
```diff
@@ -208,6 +212,69 @@ func ExampleBucket_List() {
 	// foo4.txt
 }
 
+func ExampleBucket_List_with_delimiter() {
```
Comment (zombiezen) [inline blob/example_test.go]:
```suggestion
func ExampleBucket_List_withDelimiter() {
```

File: blob/example_test.go [hunk 7]
```diff
@@ -208,6 +212,69 @@ func ExampleBucket_List() {
...
+		if err := bucket.WriteAll(ctx, key, []byte("Go Cloud"), nil); err != nil {
+			log.Fatal(err)
+		}
+	}
+
+	// list lists files in b starting with prefix. It uses the delimiter "/",
+	// and recurses into "directories", adding 2 spaces to indent each time.
+	// It will list the blobs created above because fileblob is strongly
+	// consistent, but is not guaranteed to work on all providers.
+	list := func(context.Context, *blob.Bucket, string, string) {}
```
Comment (zombiezen) [inline blob/example_test.go]: Nit:

```suggestion
	var list func(context.Context, *blob.Bucket, string, string)
```

File: blob/driver/driver.go (hunk 5 above)
Comment (vangent) [reply blob/driver/driver.go]: Done.

File: blob/example_test.go (hunk 6 above)
Comment (vangent) [reply blob/example_test.go]: Done.

File: blob/example_test.go (hunk 7 above)
Comment (vangent) [reply blob/example_test.go]: Done.

**Pull request merged**